File: bak/src/api/hanriver.go
package api

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

type hanriver struct {
	Area     string
	DateTime time.Time
	Temp     float64
	PH       float64
}

type rawhanriver struct {
	Data struct {
		Result struct {
			Code string `json:"CODE"`
		} `json:"RESULT"`
		Row []struct {
			MsrDate string `json:"MSR_DATE"`
			MsrTime string `json:"MSR_TIME"`
			SiteID  string `json:"SITE_ID"`
			WTemp   string `json:"W_TEMP"`
			WPH     string `json:"W_PH"`
			WDO     string `json:"W_DO"`
			WTN     string `json:"W_TN"`
			WTP     string `json:"W_TP"`
			WTOC    string `json:"W_TOC"`
			WPHEN   string `json:"W_PHEN"`
			WCN     string `json:"W_CN"`
		} `json:"row"`
	} `json:"WPOSInformationTime"`
}

func rawHanURL(area int) string {
	return fmt.Sprintf("http://openapi.seoul.go.kr:8088/sample/json/WPOSInformationTime/%d/%d", area, area)
}

func GetHanRiver(area int) (*hanriver, error) {
	url := rawHanURL(area)
	res, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()

	raw, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}

	var rawData rawhanriver
	err = json.Unmarshal(raw, &rawData)
	if err != nil {
		return nil, err
	}

	input := rawData.Data.Row[0]

	ys := input.MsrDate[:len(input.MsrDate)-4]
	ms := input.MsrDate[len(input.MsrDate)-4 : len(input.MsrDate)-2]
	ds := input.MsrDate[len(input.MsrDate)-2:]

	strip := strings.Split(input.MsrTime, ":")

	y, _ := strconv.ParseInt(ys, 10, 64)
	m, _ := strconv.ParseInt(ms, 10, 64)
	d, _ := strconv.ParseInt(ds, 10, 64)

	hour, _ := strconv.ParseInt(strip[0], 10, 64)
	min, _ := strconv.ParseInt(strip[1], 10, 64)

	temp, _ := strconv.ParseFloat(input.WTemp, 64)
	ph, _ := strconv.ParseFloat(input.WPH, 64)

	date := time.Date(
		int(y),
		time.Month(int(m)),
		int(d),
		int(hour),
		int(min),
		0,
		0,
		time.Now().Location(),
	)

	data := hanriver{
		Area:     input.SiteID,
		DateTime: date,
		Temp:     temp,
		PH:       ph,
	}

	return &data, nil
}

File: src/api/hanriver.rs
use std::format;
use serde::{Deserialize, Serialize};

#[derive(Serialize)]
pub struct HanRiver {
    area: String,
    datetime: String,
    temp: f64,
    ph: f64
}

#[derive(Serialize)]
pub struct HanRiverError {
    ok: i32,
    pub status: u16,
    reason: String
}

#[derive(Deserialize)]
struct RawData {
    #[serde(rename = "WPOSInformationTime")]
    data: POSInfo
}

#[derive(Deserialize)]
struct POSInfo {
    row: Vec<Row>
}

#[derive(Deserialize)]
struct Row {
    #[serde(rename = "MSR_DATE")]
    msr_date: String,
    #[serde(rename = "MSR_TIME")]
    msr_time: String,
    #[serde(rename = "SITE_ID")]
    site_id: String,
    #[serde(rename = "W_TEMP")]
    temp: String,
    #[serde(rename = "W_PH")]
    ph: String
}

fn url(area: i32) -> String {
    return format!("http://openapi.seoul.go.kr:8088/sample/json/WPOSInformationTime/{area}/{area}");
}

pub async fn get(area: i32) -> Result<HanRiver, HanRiverError> {
    if (area < 1) || (area > 5) {
        return Err(HanRiverError {
            ok: 0,
            status: 400,
            reason: "Invalid area".to_string()
        });
    }

    let res = reqwest::get(url(area)).await;
    let body = match res.unwrap().json::<RawData>().await {
        Ok(x) => x,
        Err(err) => return Err(HanRiverError {
            ok: 0,
            status: 500,
            reason: err.to_string()
        })
    };
    let mut time: Vec<String> = Vec::new();

    body.data.row[0].msr_time.split(":").for_each(|sp: &str| {
        time.push(sp.to_string());
    });

    let datetime = format!(
        "{}-{}-{}T{}:{}:00+09:00",
        &body.data.row[0].msr_date[0..4],
        &body.data.row[0].msr_date[4..6],
        &body.data.row[0].msr_date[6..8],
        time[0],
        time[1]
    );

    Ok(HanRiver {
        area: body.data.row[0].site_id.to_string(),
        datetime: datetime.to_string(),
        temp: body.data.row[0].temp.parse().unwrap_or_else(|_| 0.0),
        ph: body.data.row[0].ph.parse().unwrap_or_else(|_| 0.0)
    })
}
